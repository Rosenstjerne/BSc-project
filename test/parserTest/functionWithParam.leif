var int num1, num2;
function int printSum(int x, int y) {
  print(x + y);
}

num1 = 5;
num2 = 7;

printSum(num1, num2);
