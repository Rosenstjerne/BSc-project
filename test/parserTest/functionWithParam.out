
.data
mmapEr:
                .string "mmap failed\n" # error message string for mmap

form:
                .string "%d\n"          # form string for C printf

formTrue:
                .string "true\n"        # form string for C printf

formFalse:
                .string "false\n"       # form string for C printf

.text

.globl main

main:

                                        # CALLEE PROLOGUE
                pushq %rbp              # save caller's base pointer
                movq %rsp, %rbp         # make stack pointer new base pointer

                addq $-16, %rsp         # allocate space for local variables


                movq %rsp, %rbx         # Saves stack pointer in %rbx
                addq $-24, %rsp         # allocate space for extra intermediate
                                        # stack memory

                pushq %r8               # %r8 is callee save
                pushq %r9               # %r9 is callee save
                pushq %r10              # %r10 is callee save
                pushq %r11              # %r11 is callee save
                pushq %r12              # %r12 is callee save
                pushq %r13              # %r13 is callee save
                pushq %r14              # %r14 is callee save
                pushq %r15              # %r15 is callee save

                movq $0, -16(%rbp)      # Initializes num2 to 0
                movq $0, -8(%rbp)       # Initializes num1 to 0
                movq $5, %r13           # Moves integer into reg5
                movq %rbp, %rdx         # preparing for static link computation
                movq %r13, -8(%rdx)     # Move param reg5 into ['fun_1_main',
                                        # 'var_0_num1'] (num1)
                movq $7, %r15           # Moves integer into reg7
                movq %rbp, %rdx         # preparing for static link computation
                movq %r15, -16(%rdx)    # Move param reg7 into ['fun_1_main',
                                        # 'var_1_num2'] (num2)
                                        # CALLER PROLOGUE: empty

                movq %rbp, %rdx         # preparing for static link computation
                movq -8(%rdx), %rcx     # 
                movq %rcx, 0(%rbx)      # Move param var_0_num1 (num1) into
                                        # reg8
                movq %rbp, %rdx         # preparing for static link computation
                movq -16(%rdx), %rcx    # 
                movq %rcx, -8(%rbx)     # Move param var_1_num2 (num2) into
                                        # reg9
                pushq -8(%rbx)          # push arguement to stack
                pushq 0(%rbx)           # push arguement to stack
                pushq %rbp              # set up static link for inner function
                callq fun_2_printSum    # 
                movq %rax, -16(%rbx)    # Moves the return value from %rax into
                                        # reg10
                popq %rax               # Pops argument from stack
                popq %rax               # Pops argument from stack
                                        # CALLER EPILOGUE: empty

end_main:

                popq %r15               # restore callee save register %r15
                popq %r14               # restore callee save register %r14
                popq %r13               # restore callee save register %r13
                popq %r12               # restore callee save register %r12
                popq %r11               # restore callee save register %r11
                popq %r10               # restore callee save register %r10
                popq %r9                # restore callee save register %r9
                popq %r8                # restore callee save register %r8

                addq $24, %rsp          # deallocate stack space for extra
                                        # intermediate stack memory


                                        # CALLEE EPILOGUE
                movq %rbp, %rsp         # restore stack pointer
                popq %rbp               # restore base pointer
                ret                     # return from call

fun_2_printSum:

                                        # CALLEE PROLOGUE
                pushq %rbp              # save caller's base pointer
                movq %rsp, %rbp         # make stack pointer new base pointer

                addq $0, %rsp           # allocate space for local variables


                movq %rsp, %rbx         # Saves stack pointer in %rbx
                addq $-24, %rsp         # allocate space for extra intermediate
                                        # stack memory

                pushq %r8               # %r8 is callee save
                pushq %r9               # %r9 is callee save
                pushq %r10              # %r10 is callee save
                pushq %r11              # %r11 is callee save
                pushq %r12              # %r12 is callee save
                pushq %r13              # %r13 is callee save
                pushq %r14              # %r14 is callee save
                pushq %r15              # %r15 is callee save

                movq %rbp, %rdx         # preparing for static link computation
                movq 24(%rdx), %r8      # Move param param_2_x (x) into reg0
                movq %rbp, %rdx         # preparing for static link computation
                movq 32(%rdx), %r9      # Move param param_3_y (y) into reg1
                movq %r9, %rcx          # Moves reg1 out into %rcx for
                                        # operation
                movq %r8, %rax          # Moves reg0 out into %rax for
                                        # operation
                addq %rcx, %rax         # Operation: reg0 + reg2
                movq %rax, %r10         # Moves result into reg2
                movq %r10, %rcx         # Moves reg2 out into %rcx for print
                                        # PRINTING
                leaq form(%rip), %rdi   # pass 1. argument in %rdi
                movq %rcx, %rsi         # Moves printable object to rsi
                xorq %rax, %rax         # No floating point arguments

                callq printf@plt        # calls the printf method
                                        # END OF PRINTING
                movq $0, %r11           # Moves integer into reg3
                movq %r11, %rax         # Moves reg3 out into %rax for return
end_fun_2_printSum:

                popq %r15               # restore callee save register %r15
                popq %r14               # restore callee save register %r14
                popq %r13               # restore callee save register %r13
                popq %r12               # restore callee save register %r12
                popq %r11               # restore callee save register %r11
                popq %r10               # restore callee save register %r10
                popq %r9                # restore callee save register %r9
                popq %r8                # restore callee save register %r8

                addq $24, %rsp          # deallocate stack space for extra
                                        # intermediate stack memory


                                        # CALLEE EPILOGUE
                movq %rbp, %rsp         # restore stack pointer
                popq %rbp               # restore base pointer
                ret                     # return from call

mmap_error:
                leaq mmapEr(%rip), %rdi # pass 1. arg in %rdi
                xor %rax, %rax          # no floating point arguments
                callq printf            # calls printf method

                movq $60, %rax          # prepare for exit syscall
                xor %rdi, %rdi          # exit status 0
                syscall                 # exit


